### PR TITLE
Add support for subscriptions

### PR DESCRIPTION
**Pull request by tlvenn**
This commit adds support for subscriptions. The websocket is opened once the query is run and all headers are passed as parameters.

If the ws url or the headers are updated, the connection is closed and a new one will be opened when the query is run.

By default, it uses subscriptions-transport-ws client but an alternate subscriptionsClientBuilder can be provided.

This commit also add debouncing input values for both urls.

### CODE REVIEW

**Comment by tlvenn** [comment]
@OlegIlyenko would be great if you could review this, thanks in advance !

File: src/GraphiQLTab.jsx [hunk 1]
```diff
@@ -23,6 +24,9 @@ import {introspectionQuery} from './utility/introspectionQueries';
 
 import {buildClientSchema} from 'graphql';
 
+import {graphQLFetcher} from 'graphiql-subscriptions-fetcher';
+import {SubscriptionClient} from 'subscriptions-transport-ws';
```
Comment (theikkila) [inline src/GraphiQLTab.jsx]: This dependency isn't imported from package.json?

**Comment by theikkila** [comment]
@tlvenn hmm I couldn't find any definition for `subscriptions-transport-ws`-package from your package.json
The built bundle seems to contain old version of that library.

**Comment by tlvenn** [comment]
Ya this is on purpose, I am relying on dependency transitivity from `graphiql-subscriptions-fetcher` because `graphiql-subscriptions-fetcher` depends on a pretty old version of `subscriptions-transport-ws` and the new version has a very different API.

**Comment by tlvenn** [comment]
@OlegIlyenko any chance to review this please ?

**Comment by theikkila** [comment]
I think because subscriptions API is not yet stable we should use the latest version with graphiql-workspace and so we should also include latest library to support that new version of API.
We could make it configurable which version of API we're using but thats not obvious with current state of stability

**Comment by tlvenn** [comment]
I am all in favor of supporting the latest but I think we need to wait to the following PR to be merged until we can do that : https://github.com/apollographql/GraphiQL-Subscriptions-Fetcher/pull/9

Let's have the subs working in this project first and then it will be easy to upgrade to the latest once it's available.

**Comment by tlvenn** [comment]
@OlegIlyenko I know you must be busy and I might sound a little bit impatient but a little word / update from you to manage my expectation would go a long way. Thanks a lot in advance.

File: src/GraphiQLTab.jsx [hunk 2]
```diff
@@ -373,15 +394,51 @@ export class GraphiQLTab extends React.Component {
...
+        // port number
+        "(?::\\d{2,5})?" +
+        // resource path
+        "(?:[/?#]\\S*)?" +
+      "$", "i"
+    );
+
+    return re_url.test(url);
+  }
+
```
[the remark this answers is not in the document]
Comment (tlvenn) [reply src/GraphiQLTab.jsx]: It is surprisingly difficult to find a library that does it right, as outline here: http://mathiasbynens.be/demo/url-regex. The code is taken from Diego Perini gist from https://gist.github.com/dperini/729294 which also happens to be the foundation for url-regex lib.

A quick glance shows that while our version is very close to that lib, I have modified it a bit to be more strict around the protocol we are matching against. Our version allows only http(s) or ws(s). The lib allows any protocol.

**Comment by tlvenn** [comment]
@OlegIlyenko a bit of love please ;)

**Comment by OlegIlyenko** [comment]
Thank you very much for this great addition! 🏆 Sorry for the delay in merging the PR

**Comment by kachkaev** [comment]
@OlegIlyenko when are you planning to cut a new release? I'm keen to update [graphiql-workspace-app](https://gitlab.com/kachkaev/graphiql-workspace-app)!

**Comment by OlegIlyenko** [comment]
Also just released v1.0.5: https://github.com/OlegIlyenko/graphiql-workspace/releases/tag/v1.0.5

**Comment by kachkaev** [comment]
I'd it's 1.1.0 because of a new feature, but not a big deal if a release has already been made. Will let know when I update [graphiql-workspace-app](https://gitlab.com/kachkaev/graphiql-workspace-app)!

**Pull request merged**